File: android/network/src/main/kotlin/com/ridelink/network/control/ControlHandshake.kt
package com.ridelink.network.control

import com.ridelink.core.model.ConnTiebreak
import com.ridelink.core.model.PeerId
import com.ridelink.core.model.SessionId
import com.ridelink.core.model.SpkiHash
import com.ridelink.core.security.PeerTrust
import com.ridelink.core.security.PinDecision
import com.ridelink.core.security.TrustedPeerStore
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.util.UUID

/** Smaller `peer_id` leads (ADR-010), independent of who initiated the connection (ADR-015 A2). */
fun computeLeaderId(
    a: PeerId,
    b: PeerId,
): PeerId = if (a.value < b.value) a else b

/**
 * Placeholder session-id generation. `SessionId` (PROTOCOL §2) is documented as a ULID for
 * sortability; nothing yet parses it as one, so a random UUID string is a deliberate "smallest
 * coherent increment" stand-in (CLAUDE.md) rather than a protocol deviation. A real
 * Crockford-base32 ULID generator is a low-risk follow-up, not a phase blocker.
 */
fun freshSessionId(): SessionId = SessionId(UUID.randomUUID().toString())

sealed class HandshakeOutcome {
    data class Success(
        val remotePeerId: PeerId,
        val remoteConnTiebreak: ConnTiebreak,
        val sessionId: SessionId,
        val leaderPeerId: PeerId,
        /** Computed from the peer's TLS certificate — the only trustworthy identity input (ADR-012). */
        val peerIdentitySpkiSha256: SpkiHash,
        /**
         * Whether this peer is already trusted, needs pairing, or must be refused. Carried rather
         * than acted on here, because PROTOCOL §4.5 says pairing runs **only on the surviving
         * connection** — so the decision is made once, at handshake time, and applied after
         * duplicate-connection resolution.
         */
        val pinDecision: PinDecision,
    ) : HandshakeOutcome()

    data class Rejected(
        val errorCode: String,
    ) : HandshakeOutcome()

    object ConnectionClosed : HandshakeOutcome()
}

data class LocalHandshakeIdentity(
    val displayName: String,
    val platform: String,
    val osVersion: String,
    val appVersion: String,
    val connTiebreak: ConnTiebreak,
    /** This device's own `identity_spki_sha256` (ADR-012). Advisory on the wire; the certificate is authoritative. */
    val identitySpkiSha256: SpkiHash,
)

/**
 * PROTOCOL §4.1 HELLO/HELLO_ACK, plus the SPKI pin check that decides whether this peer is
 * trusted, unknown, or refused.
 *
 * **Every field read here comes off the wire**, so nothing in this file may throw on a malformed
 * value: a peer that sends `"peer_id": 7` or omits `conn_tiebreak` must get a clean
 * `malformed_frame` and a closed socket, not an exception that kills the coroutine running the
 * read loop. That is the same class of bug the Phase 1a hardening pass fixed for PING/PONG
 * (STATUS §2e fix 6); HELLO was not covered then because Phase 1a had no security-bearing field
 * in it to get wrong.
 *
 * **Ordering.** PROTOCOL §4.1's diagram draws the pin check before HELLO, while its normative
 * table defines the pin as "the stored pin **for that `peer_id`**" — which HELLO is what carries.
 * Both are honoured: the certificate's own structural validity is checked before HELLO is sent
 * (so an expired or unverifiable certificate never gets a session's worth of device metadata out
 * of us), and the pin comparison happens once `peer_id` is known.
 */
object ControlHandshake {
    @Suppress("ReturnCount")
    suspend fun performAsInitiator(
        socket: ControlSocket,
        localPeerId: PeerId,
        seqCounter: SeqCounter,
        monotonicNowUs: () -> Long,
        local: LocalHandshakeIdentity,
        trustedPeers: TrustedPeerStore,
    ): HandshakeOutcome {
        val security = socket.security ?: return HandshakeOutcome.Rejected(ERROR_CODE_INTERNAL)
        if (!security.peerCertificateStructurallyValid) {
            return HandshakeOutcome.Rejected(ERROR_CODE_CERTIFICATE_INVALID)
        }

        val proposal = freshSessionId()
        val sent =
            socket.writeOrClosed(
                ControlMessages.hello(
                    localPeerId = localPeerId,
                    sessionId = proposal,
                    seq = seqCounter.nextSeq(),
                    sentAtMonoUs = monotonicNowUs(),
                    displayName = local.displayName,
                    platform = local.platform,
                    osVersion = local.osVersion,
                    appVersion = local.appVersion,
                    sessionIdProposal = proposal,
                    connTiebreak = local.connTiebreak,
                    identitySpkiSha256 = local.identitySpkiSha256,
                ),
            )
        if (!sent) return HandshakeOutcome.ConnectionClosed

        val frame = socket.readFrame()
        if (frame !is FrameReadResult.Frame || frame.envelope.type != "HELLO_ACK") return mapFailure(frame)
        val payload = frame.envelope.payload

        val remotePeerId = payload.peerId("peer_id") ?: return malformed()
        val remoteConnTiebreak = payload.connTiebreak("conn_tiebreak") ?: return malformed()
        val acceptedSessionId = payload.text("accepted_session_id")?.let(::SessionId) ?: return malformed()
        val claimedLeader = payload.peerId("leader_peer_id") ?: return malformed()

        val computedLeader = computeLeaderId(localPeerId, remotePeerId)
        if (computedLeader != claimedLeader) return HandshakeOutcome.Rejected(ERROR_CODE_LEADER_MISMATCH)

        return finish(payload, security, remotePeerId, remoteConnTiebreak, acceptedSessionId, computedLeader, trustedPeers)
    }

    @Suppress("ReturnCount")
    suspend fun performAsAcceptor(
        socket: ControlSocket,
        localPeerId: PeerId,
        seqCounter: SeqCounter,
        monotonicNowUs: () -> Long,
        local: LocalHandshakeIdentity,
        trustedPeers: TrustedPeerStore,
    ): HandshakeOutcome {
        val security = socket.security ?: return HandshakeOutcome.Rejected(ERROR_CODE_INTERNAL)
        if (!security.peerCertificateStructurallyValid) {
            return HandshakeOutcome.Rejected(ERROR_CODE_CERTIFICATE_INVALID)
        }

        val frame = socket.readFrame()
        if (frame !is FrameReadResult.Frame || frame.envelope.type != "HELLO") return mapFailure(frame)
        val payload = frame.envelope.payload

        val remotePeerId = payload.peerId("peer_id") ?: return malformed()
        val remoteConnTiebreak = payload.connTiebreak("conn_tiebreak") ?: return malformed()
        val initiatorProposal = payload.text("session_id_proposal")?.let(::SessionId) ?: return malformed()

        val leader = computeLeaderId(localPeerId, remotePeerId)
        val acceptedSessionId = if (leader == localPeerId) freshSessionId() else initiatorProposal

        val sent =
            socket.writeOrClosed(
                ControlMessages.helloAck(
                    localPeerId = localPeerId,
                    sessionId = acceptedSessionId,
                    seq = seqCounter.nextSeq(),
                    sentAtMonoUs = monotonicNowUs(),
                    acceptedSessionId = acceptedSessionId,
                    connTiebreak = local.connTiebreak,
                    leaderPeerId = leader,
                    identitySpkiSha256 = local.identitySpkiSha256,
                ),
            )
        if (!sent) return HandshakeOutcome.ConnectionClosed

        return finish(payload, security, remotePeerId, remoteConnTiebreak, acceptedSessionId, leader, trustedPeers)
    }

    // One handshake's worth of already-validated fields, and one return per outcome.
    @Suppress("LongParameterList", "ReturnCount")
    private fun finish(
        payload: JsonObject,
        security: ChannelSecurity,
        remotePeerId: PeerId,
        remoteConnTiebreak: ConnTiebreak,
        sessionId: SessionId,
        leaderPeerId: PeerId,
        trustedPeers: TrustedPeerStore,
    ): HandshakeOutcome {
        // Absent is tolerated (PROTOCOL §2 rule 1: unknown/missing fields are not fatal) and simply
        // means "nothing to cross-check". Present-but-malformed is not: a peer that sends a
        // wrongly-shaped identity field is either broken or probing, and either way the session is
        // not worth having.
        val advertised = payload["identity_spki_sha256"]
        if (advertised != null && payload.spkiHash("identity_spki_sha256") == null) return malformed()

        val decision =
            PeerTrust.decide(
                storedPin = trustedPeers.byPeerId(remotePeerId)?.identitySpkiSha256,
                presentedSpki = security.peerIdentitySpkiSha256,
                helloAdvertisedSpki = payload.spkiHash("identity_spki_sha256"),
                certificateStructurallyValid = security.peerCertificateStructurallyValid,
            )
        if (decision is PinDecision.Refused) return HandshakeOutcome.Rejected(decision.code)

        return HandshakeOutcome.Success(
            remotePeerId = remotePeerId,
            remoteConnTiebreak = remoteConnTiebreak,
            sessionId = sessionId,
            leaderPeerId = leaderPeerId,
            peerIdentitySpkiSha256 = security.peerIdentitySpkiSha256,
            pinDecision = decision,
        )
    }

    /**
     * A handshake write can fail for a reason that is not this side's fault and is not an error:
     * the peer refused the certificate and closed, or the session was torn down mid-handshake. An
     * `IOException` escaping here would leave the socket unclosed and produce no outcome at all,
     * so a failed write is reported the same way a failed read already is — as
     * [HandshakeOutcome.ConnectionClosed], which the caller closes and moves on from.
     */
    @Suppress("SwallowedException") // the reason is exactly "the peer went away", which is the return value
    private suspend fun ControlSocket.writeOrClosed(envelope: com.ridelink.core.protocol.Envelope): Boolean =
        try {
            writeFrame(envelope)
            true
        } catch (io: java.io.IOException) {
            false
        }

    private fun malformed(): HandshakeOutcome = HandshakeOutcome.Rejected(ERROR_CODE_MALFORMED_FRAME)

    private fun mapFailure(frame: FrameReadResult): HandshakeOutcome =
        when (frame) {
            is FrameReadResult.ConnectionClosed -> HandshakeOutcome.ConnectionClosed
            is FrameReadResult.FrameTooLarge -> HandshakeOutcome.Rejected(ERROR_CODE_FRAME_TOO_LARGE)
            is FrameReadResult.Malformed -> HandshakeOutcome.Rejected(frame.errorCode)
            is FrameReadResult.Frame -> HandshakeOutcome.Rejected("unexpected_message_type")
        }

    /**
     * Wire-safe field readers. Each returns null — never throws — for a missing field, a field of
     * the wrong JSON type, a quoted number, or a value that fails its identifier's format check.
     * The value-class constructors (`PeerId`, `ConnTiebreak`) `require` their format because that
     * is right for values *we* produce; these wrappers are how the same types are built from
     * values a **peer** chose.
     */
    private fun JsonObject.text(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

    private fun JsonObject.peerId(key: String): PeerId? = text(key)?.takeIf { PEER_ID_FORMAT.matches(it) }?.let(::PeerId)

    private fun JsonObject.connTiebreak(key: String): ConnTiebreak? =
        text(key)?.takeIf { CONN_TIEBREAK_FORMAT.matches(it) }?.let(::ConnTiebreak)

    private fun JsonObject.spkiHash(key: String): SpkiHash? = text(key)?.let(SpkiHash::parse)

    private val PEER_ID_FORMAT = Regex("^[0-9a-f]{16}$")
    private val CONN_TIEBREAK_FORMAT = Regex("^[0-9a-f]{32}$")
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Control/ControlHandshake.swift
import Foundation
import RideLinkCore

extension JSONValue {
    var stringValue: String? {
        if case .string(let value) = self { return value }
        return nil
    }

    /// `Int64(exactly:)`, not the non-failable `Int64(_:)` — the latter **traps** (fatal error,
    /// not a thrown `Error`) on a `Double` that is NaN, infinite, or outside `Int64`'s range.
    /// PROTOCOL fields like `t1_mono_us` are attacker/peer-controlled JSON numbers, so an
    /// extreme or non-finite value here must fail safely, not crash the read loop (this
    /// session's brief §7).
    var int64Value: Int64? {
        if case .number(let value) = self { return Int64(exactly: value) }
        return nil
    }

    var boolValue: Bool? {
        if case .bool(let value) = self { return value }
        return nil
    }
}

/// Smaller `peer_id` leads (ADR-010), independent of who initiated the connection (ADR-015 A2).
func computeLeaderId(_ a: PeerId, _ b: PeerId) -> PeerId { a.value < b.value ? a : b }

/// Placeholder session-id generation. `SessionId` (PROTOCOL §2) is documented as a ULID for
/// sortability; nothing in Phase 1a parses it as one, so a random UUID string is a deliberate
/// "smallest coherent increment" stand-in (CLAUDE.md) rather than a protocol deviation.
func freshSessionId() -> SessionId { SessionId(UUID().uuidString) }

public enum HandshakeOutcome: Sendable {
    case success(remotePeerId: PeerId, remoteConnTiebreak: ConnTiebreak, sessionId: SessionId, leaderPeerId: PeerId)
    case rejected(errorCode: String)
    case connectionClosed
}

public struct LocalHandshakeIdentity: Sendable {
    public let displayName: String
    public let platform: String
    public let osVersion: String
    public let appVersion: String
    public let connTiebreak: ConnTiebreak

    public init(displayName: String, platform: String, osVersion: String, appVersion: String, connTiebreak: ConnTiebreak) {
        self.displayName = displayName
        self.platform = platform
        self.osVersion = osVersion
        self.appVersion = appVersion
        self.connTiebreak = connTiebreak
    }

    func with(connTiebreak: ConnTiebreak) -> LocalHandshakeIdentity {
        LocalHandshakeIdentity(displayName: displayName, platform: platform, osVersion: osVersion, appVersion: appVersion, connTiebreak: connTiebreak)
    }
}

/// PROTOCOL §4.1 HELLO/HELLO_ACK exchange over `PlainControlTransportPhase1a`. No TLS, no SPKI
/// pin check, no pairing (Phase 1b) — see `ProvisionalIdentity`.
public enum ControlHandshake {
    public static func performAsInitiator(
        socket: ControlConnection,
        localPeerId: PeerId,
        seqCounter: SeqCounter,
        monotonicNowUs: @Sendable () -> Int64,
        local: LocalHandshakeIdentity
    ) async throws -> HandshakeOutcome {
        let proposal = freshSessionId()
        try await socket.writeFrame(
            ControlMessages.hello(
                localPeerId: localPeerId,
                sessionId: proposal,
                seq: seqCounter.nextSeq(),
                sentAtMonoUs: monotonicNowUs(),
                displayName: local.displayName,
                platform: local.platform,
                osVersion: local.osVersion,
                appVersion: local.appVersion,
                sessionIdProposal: proposal,
                connTiebreak: local.connTiebreak
            )
        )

        let frame = await socket.readFrame()
        guard case .frame(let envelope, _) = frame, envelope.type == "HELLO_ACK" else { return mapFailure(frame) }
        let payload = envelope.payload

        guard
            let remotePeerIdString = payload["peer_id"]?.stringValue,
            let remoteConnTiebreakString = payload["conn_tiebreak"]?.stringValue,
            let acceptedSessionIdString = payload["accepted_session_id"]?.stringValue,
            let claimedLeaderString = payload["leader_peer_id"]?.stringValue
        else {
            return .rejected(errorCode: errorCodeMalformedFrame)
        }

        let remotePeerId = PeerId(remotePeerIdString)
        let claimedLeader = PeerId(claimedLeaderString)
        let computedLeader = computeLeaderId(localPeerId, remotePeerId)
        guard computedLeader.value == claimedLeader.value else { return .rejected(errorCode: errorCodeLeaderMismatch) }

        return .success(
            remotePeerId: remotePeerId,
            remoteConnTiebreak: ConnTiebreak(remoteConnTiebreakString),
            sessionId: SessionId(acceptedSessionIdString),
            leaderPeerId: computedLeader
        )
    }

    public static func performAsAcceptor(
        socket: ControlConnection,
        localPeerId: PeerId,
        seqCounter: SeqCounter,
        monotonicNowUs: @Sendable () -> Int64,
        local: LocalHandshakeIdentity
    ) async throws -> HandshakeOutcome {
        let frame = await socket.readFrame()
        guard case .frame(let envelope, _) = frame, envelope.type == "HELLO" else { return mapFailure(frame) }
        let payload = envelope.payload

        guard
            let remotePeerIdString = payload["peer_id"]?.stringValue,
            let remoteConnTiebreakString = payload["conn_tiebreak"]?.stringValue,
            let initiatorProposalString = payload["session_id_proposal"]?.stringValue
        else {
            return .rejected(errorCode: errorCodeMalformedFrame)
        }

        let remotePeerId = PeerId(remotePeerIdString)
        let initiatorProposal = SessionId(initiatorProposalString)
        let leader = computeLeaderId(localPeerId, remotePeerId)
        let acceptedSessionId = leader.value == localPeerId.value ? freshSessionId() : initiatorProposal

        try await socket.writeFrame(
            ControlMessages.helloAck(
                localPeerId: localPeerId,
                sessionId: acceptedSessionId,
                seq: seqCounter.nextSeq(),
                sentAtMonoUs: monotonicNowUs(),
                acceptedSessionId: acceptedSessionId,
                connTiebreak: local.connTiebreak,
                leaderPeerId: leader
            )
        )

        return .success(
            remotePeerId: remotePeerId,
            remoteConnTiebreak: ConnTiebreak(remoteConnTiebreakString),
            sessionId: acceptedSessionId,
            leaderPeerId: leader
        )
    }

    private static func mapFailure(_ frame: FrameReadResult) -> HandshakeOutcome {
        switch frame {
        case .connectionClosed: .connectionClosed
        case .frameTooLarge: .rejected(errorCode: errorCodeFrameTooLarge)
        case .malformed(let code): .rejected(errorCode: code)
        case .frame: .rejected(errorCode: "unexpected_message_type")
        }
    }
}
